### PR TITLE
[FEATURE] Create unit test for rust code

### LINKED ISSUE #911: [PROPOSAL] Add some Unit Test for CCExtractor
Please prefix your issue with one of the following: [BUG], [PROPOSAL], [QUESTION].

CCExtractor version (using the --version parameter preferably) : **X.X**

**In raising this issue, I confirm the following (please check boxes, eg [X]):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [x] I have checked that the issue I'm posting isn't already reported.
- [x] I have checked that the issue I'm porting isn't already solved and no duplicates exist in [closed issues](https://github.com/CCExtractor/ccextractor/issues?q=is%3Aissue+is%3Aclosed) and in [opened issues](https://github.com/CCExtractor/ccextractor/issues)
- [x] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.
- [x] I have used the latest available version of CCExtractor to verify this issue exists.

**My familiarity with the project is as follows (check one, eg [X]):**

- [ ] I have never used CCExtractor.
- [x] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [ ] I am an active contributor to CCExtractor.

**Necessary information**
- Is this a regression (did it work before)? [ ] NO | [ ] YES - *please specify the last known working version*
- What platform did you use? [ ] Windows - [ ] Linux - [ ] Mac
- What where the used arguments? `-autoprogram`

**Video links**

Please make the affected input file available for us (no screenshots, those don't help!). Public links to Dropbox, Google Drive, etc, are all fine. If it is not possible to make it available publicly, send us a private invitation (both Dropbox and Google Drive allow that). In this case we will download the file and upload it to the private developer repository.

Do *not* upload your file to any location that will require us to sign up or endure a wait list, slow downloads, etc. If your upload expires make sure you keep it active somehow (replace links if needed). Keep in mind that while we go over all tickets some may take a few days, and it's important we have the file available when we actually need it.

**Additional information**
We need to add some unit test to CCExtractor to make sure things run properly (as the latest unit test commit is a year ago)

PS: Make sure you set an alert in GitHub so you get notifications about your ticket. We may need to ask questions and we do everything inside GitHub's system.

### PR DESCRIPTION
**Pull request by IshanGrover2004**
**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I have considered, and confirmed that this submission will be valuable to others.
- [x] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [x] I give this submission freely, and claim no ownership to its content.
- [x] **I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).**

**My familiarity with the project is as follows (check one):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [x] I am an active contributor to CCExtractor.

---

Creating Unit test cases for rust modules(`decoder` + `hardsubx`).
Closes #911

### CODE REVIEW

**Comment by IshanGrover2004** [comment]
@PunitLodha

**Comment by IshanGrover2004** [comment]
The decoder moulde's unit test  done @PunitLodha

**Comment by PunitLodha** [comment]
@IshanGrover2004 Can you also add a github CI check for running the unit tests

**Comment by IshanGrover2004** [comment]
Sure @PunitLodha

**Comment by IshanGrover2004** [comment]
Don't merge this right now, I am adding test cases & CI check for new `lib_ccxr` part

**Comment by IshanGrover2004** [comment]
@PunitLodha You can proceed for merging. I have added CI + all modules unit test cases

**Review by PunitLodha — APPROVED** [review]

**Comment by PunitLodha** [comment]
If you find time you can write the extra comments, otherwise it's fine

File: src/rust/src/decoder/window.rs [hunk 1]
```diff
@@ -548,3 +533,224 @@ impl Default for dtvcc_pen_attribs {
...
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{cb_708, cb_field1, cb_field2};
+
+    #[test]
+    fn test_update_time_show() {
+        set_temp_cb_values();
```
[the remark this answers is not in the document]
Comment (IshanGrover2004) [reply src/rust/src/decoder/window.rs:566]: Yeah sure

File: .github/workflows/format.yml [hunk 2]
```diff
@@ -52,3 +52,29 @@ jobs:
...
+      - uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          override: true
+      - name: Build and test main module
+        run: cargo test
+        working-directory: ./src/rust
+      - name: Build and test lib_ccxr
+        run: cargo test
+        working-directory: ./src/rust/lib_ccxr
```
[the remark this answers is not in the document]
Comment (IshanGrover2004) [reply .github/workflows/format.yml:80]: @prateekmedia #1619 is merged, but you forgot to test code within newly formed module `lib_ccxr`

File: src/rust/src/decoder/mod.rs [hunk 3]
```diff
@@ -224,3 +224,99 @@ impl Default for dtvcc_symbol {
...
+        let mut dtvcc_ctx = get_zero_allocated_obj::<dtvcc_ctx>();
+        let mut decoder = Dtvcc::new(&mut dtvcc_ctx);
+
+        // Case 1:  cc_type = 2
+        let mut dtvcc_report = ccx_decoder_dtvcc_report::default();
+        decoder.report = &mut dtvcc_report;
+        decoder.is_header_parsed = true;
+        decoder.is_active = true;
+        decoder.report_enabled = true;
+        decoder.packet = vec![0xC2, 0x23, 0x45, 0x67, 0x00, 0x00];
```
[the remark this answers is not in the document]
Comment (IshanGrover2004) [reply src/rust/src/decoder/mod.rs:251]: Not possible tbh, bcs these are based on random values resulting of manipulation done in the functions & parameters passed as inputs. On based on that the resulted values are tested, I did not planned for any particular data.

**Comment by IshanGrover2004** [comment]
@PunitLodha Added the CI & fixed the warnings as well

**Review by PunitLodha — APPROVED** [review]

**Pull request merged**